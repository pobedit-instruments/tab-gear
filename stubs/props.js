import { bindActionCreators } from 'redux';

import actual from './tabs';
import * as Actions from '../views/actions';
import * as ActionTypes from '../views/constants/ActionTypes';
import config from '../config';

export const tabs = {
	actual,
	masked: false,
	active: actual.find(tab => tab.active),
	search: actual
};

export const type = ActionTypes.SHOW_TABS;

export const actions = new function () {
	let actions = {};

	for (let [name, value] of Object.entries(Actions)) {
		actions[name] = bindActionCreators(value, () => {});
	}

	return actions;
};

export { config };
