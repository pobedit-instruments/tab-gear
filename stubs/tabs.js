export default [
	{
		title : 'Google Cloud Computing, Hosting Services & APIs | Google Cloud Platform',
		url   : 'https://cloud.google.com',
		favIconUrl: 'https://cloud.google.com/_static/cc7442d8c7' +
					'/images/cloud/icons/favicons/favicon.ico',
		id : 0,
		index : 0,
		active: true,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Google Plus +++++++++++++++++++++++++++++++++++++++++++',
		url   : 'https://developers.google.com/+/',
		favIconUrl: 'https://developers.google.com/site-assets/logo-google-plus-white.svg',
		id : 1,
		index : 1,
		active: false,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Google Play',
		url   : 'https://play.google.com',
		favIconUrl: 'https://developers.google.com/_static/cc7442d8c7/images/favicon.png',
		id : 2,
		index : 2,
		active: false,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Google Maps',
		url   : 'https://maps.google.com',
		favIconUrl: 'https://www.google.com/images/branding/product/ico/maps_32dp.ico',
		id : 3,
		index : 3,
		active: false,
		highlighted: false,
		incognito: true
	},

	{
		title : 'Google Translate',
		url   : 'https://translate.google.ru/',
		favIconUrl: 'https://translate.google.ru/favicon.ico',
		id : 4,
		index : 4,
		active: false,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Android',
		url   : 'https://developer.android.com',
		favIconUrl: 'https://developer.android.com/favicon.ico',
		id : 5,
		index : 5,
		active: false,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Google Chrome',
		url   : 'https://google.com/chrome',
		favIconUrl: 'https://www.google.com/images/icons/product/chrome-32.png',
		id : 6,
		index : 6,
		highlighted: true,
		active: false
	},

	{
		title : 'Google Developers',
		url   : 'https://developers.google.com/web/',
		favIconUrl: 'https://developers.google.com/web/images/web-fundamentals-icon192x192.png',
		id : 7,
		index : 7,
		active: false,
		highlighted: false,
		incognito: false
	},

	{
		title : 'Google Firebase',
		url   : 'https://firebase.google.com/',
		favIconUrl: 'https://firebase.google.com/_static/cc7442d8c7/images/firebase/favicon.png',
		id : 9,
		index : 9,
		highlighted: false,
		active: false
	}
];
