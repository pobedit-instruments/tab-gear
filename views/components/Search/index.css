@import "../decor.css";

.tg-search {
	position: relative;

	&.is-active {
		.tg-search__input {
			z-index: 0;
		}

		.tg-search__clear {
			opacity: 1;
			z-index: 2;
		}
	}
}

	.tg-search__input {
		color: var(--search-color);
		font-size: var(--search-font-size);
		line-height: var(--search-line-height);
		padding: var(--padding);
		border: none;
		outline: none;
		position: relative;
		transition: box-shadow .4s cubic-bezier(.4, 0, 1, 1);
		width: 100%;
		z-index: 1;

		&:focus {
			box-shadow: 0 -12px 15px -16px var(--search-box-focus-color);
		}
	}

	.tg-search__clear {
		transition: opacity .3s ease-in;
		opacity: 0;
		z-index: 0;

		&:after {
			content: 'close';
			color: var(--search-clear-color);
			cursor: pointer;
			font-size: var(--close-glyph-font-size);
			font-family: 'Material';
			line-height: var(--search-clear-line-height);
			position: absolute;
			right: var(--padding);
			top: 19px;
			-webkit-font-feature-settings: 'liga';
			-webkit-font-smoothing: antialiased;
			/*z-index: 1;*/
		}

		&:hover:after {
			color: var(--search-color);
		}
	}
