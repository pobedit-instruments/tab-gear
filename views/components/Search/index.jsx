import React, { Component, PropTypes } from 'react';
import * as ActionTypes from '../../constants/ActionTypes';
import BEMHelper from 'react-bem-helper';
import { bind } from 'decko';
import { KeyKit } from 'keykit';

import './index.css';

class Search extends Component {
	constructor () {
		super(...arguments);

		this.class = new BEMHelper({
			name: 'tg-search'
		});

		this.state = {
			value: '',
			switchTab: false
		};
	}

	shouldComponentUpdate () {
		let types = [
			ActionTypes.DISABLE_TABS,
			ActionTypes.SEARCH_TABS
		];

		return types.includes(this.props.type);
	}

	componentDidMount () {
		document.addEventListener('keydown', this.onKeyDown);
	}

	componentWillUnmount () {
		document.removeEventListener('keydown', this.onKeyDown);
	}

	@bind
	onKeyDown (event) {
		switch (event.key) {
			case 'Tab':
				return event.preventDefault();

			case 'Escape':
				return this.input.blur();

			default:
				try {
					let { visible, printable } = KeyKit.key(event.keyCode);

					if (visible && printable) {
						 this.input.focus();
					}
				}
				catch (error) {
					console.error(error);
				}
		}
	}

	@bind
	onFocus (event) {
		let { actions, tabs } = this.props;

		let state = tabs.actual.length !== tabs.search.length;

		actions.Tabs.maskTabs(tabs.actual, state);
	}

	@bind
	onBlur (event) {
		let { actions, tabs, type } = this.props;

		// searchTabs event should be called before.
		// Also we shouldn't care about "focusin" event
		if (!event.relatedTarget || event.relatedTarget.dataset.name !== 'tab') {
			actions.Tabs.maskTabs(tabs.actual, false);
		}
	}

	@bind
	clearSearch (event) {
		let { tabs, actions } = this.props;

		actions.Tabs.searchTabs(tabs.actual);
		this.setState({ value: '' });
	}

	@bind
	searchTabs (event) {
		let { tabs, actions } = this.props;
		let { value } = event.target;

		this.setState({ value });

		actions.Tabs.searchTabs(tabs.actual, value);
		event.stopPropagation();
	}

	render () {
		let { value } = this.state;

		let state = {
			extra: {
				'is-active': value.length > 0
			}
		};

		return <div { ...this.class(state) }>
					<input className="tg-search__input" type="text"
					       value={ value }
					       ref={ input => this.input = input }
					       placeholder={ chrome.i18n.getMessage('search') }
					       onFocus={ this.onFocus }
					       onBlur={ this.onBlur}
					       onInput={ this.searchTabs } />

				<div className="tg-icon tg-search__clear" onClick={ this.clearSearch } />
			</div>;
	}
}

Search.propTypes = {
	tabs   : PropTypes.object.isRequired,
	actions: PropTypes.object.isRequired
};

export default Search;
