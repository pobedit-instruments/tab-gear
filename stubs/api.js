import chrome from 'chrome-stub';

import tabs from './tabs';
import locale from '../locale/en/messages.json';

chrome.i18n._locales = locale;

let methods = [
	'query',
	'remove',
	'move',
	'create',
	'discard',
	'highlight'
];

for (let method of methods) {
	chrome.tabs[method].yields(tabs);
}

export default chrome;
