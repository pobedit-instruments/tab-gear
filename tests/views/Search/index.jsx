import React from 'react';
import renderer from 'react-test-renderer';
// import { shallow, mount, render } from 'enzyme';

import Search from '../../../views/components/Search';
import * as props from '../../../stubs/props';

test('Search view', () => {
	let component = renderer.create(<Search { ...props } />);
	let view = component.toJSON();

	let actual = expect(view);

	actual.toMatchSnapshot();
});
