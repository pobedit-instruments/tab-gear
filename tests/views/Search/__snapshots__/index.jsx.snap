exports[`test Search view 1`] = `
<div
  className="tg-search">
  <input
    className="tg-search__input"
    onBlur={[Function]}
    onFocus={[Function]}
    onInput={[Function]}
    placeholder="Searchâ€¦"
    type="text"
    value="" />
  <div
    className="tg-icon tg-search__clear"
    onClick={[Function]} />
</div>
`;
